Support narrowing down area queries with species parameters

corviday opened the pull request:
We would like users of the front end to be able to find the watershed they care about by specifying which salmon species they are interested in. To this end, we need the `region` query in the back end to accept parameters for `common_name` and optionally `subgroup` and only return regions in which salmon of that species and subgroup are present.

Questions for reviewers:
* the database query is constructed differently depending on parameters received, which might be confusing. Have I [explained the logic adequately in comments](https://github.com/pacificclimate/scip-backend/blob/8e96cf23ebc83c19e4001b4f8f874bdf61ca7f62/scip/api/region.py#L39)?
* for parameter verification, there are now hardcoded lists of valid region kinds, salmon species and salmon species subgroups in [validators.py](https://github.com/pacificclimate/scip-backend/blob/8e96cf23ebc83c19e4001b4f8f874bdf61ca7f62/scip/api/validators.py). I chose to do it this way for speed, but perhaps it would be better to fetch these lists from the database. Opinions?

Resolves #6

rod-glover reviewed (commented):
Wow, great work, Lee. 

I mostly have modest code improvements to suggest, but I also have a question about how that `func.distinct` affects the queries, and I'd love to know if you know.

corviday commented on scip/api/region.py:
```diff
@@ -22,27 +27,60 @@ def region(session, kind, overlap=None, name=None, code=None):
 
+                if subgroup:
+                    q = q.filter(Taxon.subgroup == subgroup)
+                q = q.distinct(ConservationUnit.name)
```
I don't know its exact effect on the generated SQL - is there a way to see the generated SQL?

I do know that without it we get thousands and thousands of query results, which I believe is caused by returning a query result for every intersection of a watershed and a conservation unit. That is, if the "Fraser Canyon" watershed intersects six different conservation units where Chum salmon are found, we would get a "Fraser River" query result for each of those interactions, which is obviously not the desired behavior.

rod-glover replied:
Yeah, that makes sense to me, but I would think it had to specify the whole row (all columns) returned by the query, not just one of the columns. But maybe SQLAlchemy is smart about that.

You can look at the SQL generated by query just by printing `str(q)`.

corviday replied:
Huh, SQLAlchemy is apparently using the postgres-specific SELECT DISTINCT ON here, which seems to behave as I want it to. Guess I got lucky.

```sql
SELECT DISTINCT ON (salmon_geometry.conservation_unit.name) salmon_geometry.conservation_unit.name AS name, salmon_geometry.conservation_unit.code AS code, ST_AsGeoJSON(salmon_geometry.conservation_unit.boundary) AS boundary, ST_AsGeoJSON(salmon_geometry.conservation_unit.outlet) AS outlet 
FROM salmon_geometry.conservation_unit JOIN salmon_geometry.population ON salmon_geometry.population.conservation_unit_id = salmon_geometry.conservation_unit.conservation_unit_id JOIN salmon_geometry.taxon ON salmon_geometry.population.taxon_id = salmon_geometry.taxon.taxon_id 
WHERE salmon_geometry.taxon.common_name = %(common_name_1)s AND salmon_geometry.taxon.subgroup = %(subgroup_1)s
```

rod-glover replied:
That is extremely cool. I didn't know about SELECT DISTINCT ON. I think you might do even better to reduce it to just SELECT DISTINCT:

> SELECT DISTINCT eliminates duplicate rows from the result. SELECT DISTINCT ON eliminates rows that match on all the specified expressions. SELECT ALL (the default) will return all candidate rows, including duplicates. (See [DISTINCT Clause](https://www.postgresql.org/docs/current/sql-select.html#SQL-DISTINCT) below.)

I bet that would be `q = q.distinct()`.

corviday commented on scip/api/region.py:
```diff
@@ -22,27 +27,60 @@ def region(session, kind, overlap=None, name=None, code=None):
+                        ConservationUnit.code.label("code"),
+                        func.ST_AsGeoJSON(ConservationUnit.boundary).label("boundary"),
+                        func.ST_ASGeoJSON(ConservationUnit.outlet).label("outlet"),
+                    )
```
I didn't think sqlalchemy worked like that, but maybe I'm wrong. I'll give it a shot!

rod-glover replied:
I just meant to start with the `q = session.query(...)` part (which establishes the columns selected) and then keep augmenting it as you do elsewhere.

rod-glover commented:
To answer the questions you posed: 

Comments: I think so. But as I noted in a comment, I think the code for `region` is too long and complicated for simple reading, and that it would be better to split the middle of it into two query-generating functions with one level less of if statement. Then you can put the appropriate comments explaining each type of query into those functions.

Validators: Definitely fetch those from the database, or from a config file. If the info is already in the db, then that's a natural. One always regrets hardcoded values in the end.

corviday commented:
Changes:

- Refactored query generation in `region` API. 
- Parameter verification functions now run against the database
- Added new "taxon" API that does nothing but list all the species in the database - useful for parameter verification, and also for species selection in the front end

corviday commented:
Demo:

[New taxon API shows all known species](https://services.pacificclimate.org/dev/scip/api/taxon)

[All watersheds (245)](https://services.pacificclimate.org/dev/scip/api/region?kind=watershed)

[Watersheds with Pink Salmon in them (117)](https://services.pacificclimate.org/dev/scip/api/region?kind=watershed&common_name=Pink)

[Watersheds with Even-Year Pink Salmon in them (37)](https://services.pacificclimate.org/dev/scip/api/region?kind=watershed&common_name=Pink&subgroup=Even)

rod-glover commented on scip/api/taxon.py:
```diff
@@ -0,0 +1,31 @@
+    results = q.all()
+
+    result_list = [
+        {
```
This can be done more compactly, and without having to name the attributes explicitly, with `result._mapping`. See https://docs.sqlalchemy.org/en/20/core/connections.html#sqlalchemy.engine.Row._mapping

rod-glover reviewed (approved):
Looks great, Lee!
